lib/route.js

```diff
@@ -208,117 +201,132 @@ function buildRouting (options) {
       }
 
       if (prefixing === false) {
-      // run 'onRoute' hooks
+        // run 'onRoute' hooks
         for (const hook of this[kHooks].onRoute) {
-          try {
-            hook.call(this, opts)
-          } catch (error) {
-            done(error)
-            return
-          }
+          hook.call(this, opts)
         }
       }
 
+      const constraints = opts.constraints || {}
       const config = {
         ...opts.config,
         url,
         method: opts.method
       }
-      const constraints = opts.constraints || {}
-      if (opts.version) {
-        warning.emit('FSTDEP006')
-        constraints.version = opts.version
-      }
-
       const context = new Context(
         opts.schema,
         opts.handler.bind(this),
         this[kReply],
         this[kRequest],
         this[kContentTypeParser],
-        config,
-        opts.errorHandler || this[kErrorHandler],
-        opts.bodyLimit,
-        opts.logLevel,
-        opts.logSerializers,
-        opts.attachValidation,
-        this[kReplySerializerDefault],
-        opts.schemaErrorFormatter || this[kSchemaErrorFormatter]
+        config
       )
 
-      const headRouteExists = router.find('HEAD', path) != null
-
-      try {
-        router.on(opts.method, opts.url, { constraints }, routeHandler, context)
-      } catch (err) {
-        done(err)
-        return
+      if (opts.version) {
+        warning.emit('FSTDEP006')
+        constraints.version = opts.version
       }
 
-      const { exposeHeadRoute } = opts
-      const hasRouteExposeHeadRouteFlag = exposeHeadRoute != null
-      const shouldExposeHead = hasRouteExposeHeadRouteFlag ? exposeHeadRoute : globalExposeHeadRoutes
+      const headRouteExists = opts.method === 'HEAD' && router.find('HEAD', url, constraints) != null
+
+      // TODO: improve validation for auto expose HEAD route before
+      // adding a new route
+
+      // There's a bug when ignoreTrailingSlash !== true
+      // See
+      /*
+      └── /prefix
+          └── / (GET)
+              / (HEAD)
+              └── / (GET)
+      **/
+
+      if (!headRouteExists) {
+        try {
+          router.on(opts.method, opts.url, { constraints }, routeHandler, context)
+        } catch (error) {
+          const isDuplicatedRoute = Array.isArray(error.message.match(`Method '${opts.method}' already declared for route '${opts.url}'`))
```

Comment:
the error instance is not proto of `FST_ERR_DUPLICATED_ROUTE` so you can just check `if (error instanceof FST_ERR_DUPLICATED_ROUTE)`?

Reply:
perhaps it's thrown directly by `find-my-way` so you can just ignore it.

Reply:
Yes, we're relying on `find-my-way`  throwing an error and just decorating it on top. I think the way I'm doing the validation to see if the error message matches with the error we're expecting might not be the best, I'm open to suggestions 🙂 

Regarding ignoring `find-my-way` errors, I'm not so sure. I think is possible that there are validation issues thrown by `find-my-way` and we shouldn't silence it and be handled by the implementation. What do you think?

Reply:
What I mean with _"ignore it"_ was ignore my comment if it's throw by `find-my-way`, I'm ok with the currently approach so far :smile:.